File: 2-dockerize_nestjs/libs/common/src/geo.service.ts
import { Injectable } from '@nestjs/common';
import { IGeoService, GeoLocation } from '@mytravels/contract';
import * as exifr from 'exifr';

@Injectable()
export class GeoService implements IGeoService {
  async extractGeoLocation(buffer: Buffer): Promise<GeoLocation> {
    try {
      const data = await exifr.parse(buffer, { gps: true });
      if (data?.latitude && data?.longitude) {
        return { latitude: data.latitude, longitude: data.longitude };
      }
    } catch {
      // no GPS data in image
    }
    return { latitude: 0, longitude: 0 };
  }
}

File: 0-preview/common/mytravels.common/Services/GeoService.cs
using MetadataExtractor;
using MetadataExtractor.Formats.Exif;
using mytravels.contract.Interfaces;

namespace mytravels.common.Services
{
    public class GeoService : IGeoService
    {
        public GeoLocation ExtractGeoLocation(Stream stream)
        {
            var directories = ImageMetadataReader.ReadMetadata(stream);
            // Find the GPS directory
            var gpsDirectory = directories.OfType<GpsDirectory>().FirstOrDefault();
            if (gpsDirectory is null)
            {
                return new GeoLocation(0, 0);
            }
            else
            {
                // Get the latitude and longitude
                GeoLocation geoLocation = gpsDirectory.GetGeoLocation();
                if (geoLocation is null)
                {
                    return new GeoLocation(0, 0);
                }
                return geoLocation;
            }
        }
    }
}
